### PR TITLE
Module 4/exercise

### PR DESCRIPTION
**Pull request by dagoud**
# Summary

Module 4 exercise - completed with stretch goals:

- [x] Your README documents the command to provision a VM from an Ansible Control Node 

Git repository contains:
- [x] Inventory file 
- [x] Playbook file 
- [x] .env.j2 
- [x] todoapp.service

## Stretch Goals
- [x] Install via Gunicorn
- [x] Set environment File via Gunicorn in todoapp.service
- [x] You can use the "bind" option of gunicorn to specify a host and port number.
- [x] Configure the todoapp.service to require todoapp.socket
- [x] Write log events from gunicorn to a file on the host machine, so you can easily view them. 

## Limiting Permissions
- [x] Use Ansible to create a new "todoapp" user (made so can create new user via input from env variable)
- [x] Make sure that user can use Poetry
- [x] Update your systemd service file to run your service as the new user

### CODE REVIEW

File: gunicorn.conf.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+
+# bind = "0.0.0.0:80"
+# raw_env = ["SECRET_KEY=secret-key"]
```
Comment (JackMeadDev) [inline gunicorn.conf.py]: This looks pretty unused - worth deleting this file?

File: my-ansible-managed-playbook.yml [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+      default: trello_token
+    
+    - name: user_name
+      prompt: What is your user name?
+      private: no
+      default: todoapp_user
+
+  tasks:
+  - name: Create user {{ user_name }} in ~{{ user_name }}
+    ansible.builtin.user:
```
Comment (JackMeadDev) [inline my-ansible-managed-playbook.yml:27]: Nice work adding some more stages to this!

File: my-ansible-managed-playbook.yml [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    become_user: "{{ user_name }}"
+  
+  - debug:  
+      var: poetry_version.stdout_lines
+
+  - name: Create todoapp directory
+    ansible.builtin.file:
+      path: /opt/todoapp
+      state: directory
+      mode: '477'
```
Comment (JackMeadDev) [inline my-ansible-managed-playbook.yml:91]: This seems an odd permission setting! Do you really want the owner to have less permission than an average user? 
744 is a more common setting

File: my-ansible-managed-playbook.yml [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+      dest: /opt/todoapp
+      version: module-4/exercise
+    register: todoapp_git_checkout
+    become: yes
+
+  - debug:  
+      var: todoapp_git_checkout.stdout_lines
+
+  - name : Change dir to todoapp on Managed Node
+    ansible.builtin.shell: git status
```
Comment (JackMeadDev) [inline my-ansible-managed-playbook.yml]: I don't think the name matches the action here

File: my-ansible-managed-playbook.yml [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+  - debug:  
+      var: todoapp_install.stdout_lines
+
+  - name: Template a .env file to /opt/todoapp/.env
+    ansible.builtin.template:
+      src: ~/.env.j2
+      dest: /opt/todoapp/.env
+  
+  - name : Print .env contents to confirm written variables correctly
```
Comment (JackMeadDev) [inline my-ansible-managed-playbook.yml:138]: Do you actually want this with your secret values?

**Review by JackMeadDev — APPROVED** [review]
Nice work, as you'd hinted to me your playbook looks much more sophisticated than the requirements of this project so well done on pushing that further! I've left a few optional thoughts below but it looks ready to merge as is if you're happy. 😄 

By default, I'll clear out the VMs at the end of next week unless I hear otherwise as we won't be using them again, so do rescue anything you might need from them or let me know if that's an issue

File: my-ansible-managed-playbook.yml (hunk 2 above)
Comment (dagoud) [reply my-ansible-managed-playbook.yml:27]: Thanks : )

File: my-ansible-managed-playbook.yml (hunk 5 above)
Comment (dagoud) [reply my-ansible-managed-playbook.yml:138]: I thought this was how it said to do it in the exercise - no secret values are actually in there - the secrets are passed in when running the ansible script.

```
# Flask server configuration.
FLASK_APP=todo_app/app
FLASK_ENV={{ flask_env }}

# Change the following values for local development.
SECRET_KEY={{ secret_key }}
TRELL_API_TOKEN={{ trello_token }}
```

Is this ok?

File: my-ansible-managed-playbook.yml (hunk 3 above)
Comment (dagoud) [reply my-ansible-managed-playbook.yml:91]: In the exercise it said: 

Make sure your "todoapp" user cannot write to the /opt/todoapp directory (they will still need to be able to read from it!) and does not have any other unnecessary
permissions - unlike ec2-user they should not be able to use sudo.

So this is what I was aiming for - I thought 744 would give the user write and execute permissions as well, and others only read? That is why I lowered the user to 4. Was this incorrect?

File: my-ansible-managed-playbook.yml (hunk 4 above)
Comment (dagoud) [reply my-ansible-managed-playbook.yml]: Updated thanks

**Pull request merged**